File: angular/src/models/product.ts
import { Category } from './category';
import { Supplier } from './supplier';
import { Observable } from 'rxjs';

export class Product {
  private id: number;
  private productName: string;
  private category: {};
  private fullPrice: number;
  private salePrice: number;
  private discount: number;
  private availability: boolean;
  private supplier: {};

  constructor(id: number, 
     productName: string, 
     category: {}, 
     fullPrice: number, 
     salePrice: number, 
     discount: number, 
     availability: boolean, 
     supplier: {}) {
    this.id = id;
    this.productName = productName;
    this.category = category;
    this.fullPrice = fullPrice;
    this.salePrice = salePrice;
    this.discount = discount;
    this.availability = availability;
    this.supplier = supplier;
  }


    /**
     * Getter $id
     * @return {number}
     */
	public get $id(): number {
		return this.id;
	}

    /**
     * Getter $productName
     * @return {string}
     */
	public get $productName(): string {
		return this.productName;
	}

    /**
     * Getter $category
     * @return {Category}
     */
	public get $category(): {} {
		return this.category;
	}

    /**
     * Getter $fullPrice
     * @return {number}
     */
	public get $fullPrice(): number {
		return this.fullPrice;
	}

    /**
     * Getter $salePrice
     * @return {number}
     */
    public get $salePrice(): number {
     return this.salePrice;
}

    /**
     * Getter $discount
     * @return {number}
     */
    public get $discount(): number {
     return this.discount;
}

    /**
     * Getter $availability
     * @return {boolean}
     */
	public get $availability(): boolean {
		return this.availability;
	}

    /**
     * Getter $supplier
     * @return {Supplier}
     */
	public get $supplier(): {} {
		return this.supplier;
	}

    /**
     * Setter $id
     * @param {number} value
     */
	public set $id(value: number) {
		this.id = value;
	}

    /**
     * Setter $productName
     * @param {string} value
     */
	public set $productName(value: string) {
		this.productName = value;
	}

    /**
     * Setter $category
     * @param {Category} value
     */
	public set $category(value: {}) {
		this.category = value;
	}

    /**
     * Setter $fullPrice
     * @param {number} value
     */
	public set $fullPrice(value: number) {
		this.fullPrice = value;
	}

    /**
     * Setter $salePrice
     * @param {number} value
     */
    public set $salePrice(value: number) {
     this.salePrice = value;
}

    /**
     * Setter $discount
     * @param {number} value
     */
    public set $discount(value: number) {
     this.discount = value;
}

    /**
     * Setter $availability
     * @param {boolean} value
     */
	public set $availability(value: boolean) {
		this.availability = value;
	}

    /**
     * Setter $supplier
     * @param {Supplier} value
     */
	public set $supplier(value: {}) {
		this.supplier = value;
	}

}

File: src/main/java/com/tts/ttsdashboardproject/entities/Product.java
package com.tts.ttsdashboardproject.entities;

import org.hibernate.annotations.Formula;
import javax.persistence.*;
import java.util.Objects;

@Entity
@Table(name = "products")
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private int id;

    @Column(name = "productname")
    private String productName;

    @ManyToOne
    @OrderBy("categoryId")
    @JoinColumn(name="category", referencedColumnName = "categoryid")
    private Category category;

    @Column(name = "fullprice")
    private double fullPrice;

    @Column(name = "saleprice")
    private double salePrice;

    @Formula("(fullprice - saleprice) / fullprice")
    private Double discount;

    @Column(name = "availability")
    private boolean availability;

    @ManyToOne
    @JoinColumn(name = "supplier", referencedColumnName = "supplierid")
    private Supplier supplier;

    public Product() {}

    public Product(String productName, Category category, double fullPrice, double salePrice, boolean availability, Supplier supplier) {
        this.productName = productName;
        this.category = category;
        this.fullPrice = fullPrice;
        this.salePrice = salePrice;
        this.availability = availability;
        this.supplier = supplier;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getProductName() {
        return productName;
    }

    public void setProductName(String productName) {
        this.productName = productName;
    }

    public Category getCategory() {
        return category;
    }

    public void setCategory(Category category) {
        this.category = category;
    }

    public double getFullPrice() {
        return fullPrice;
    }

    public void setFullPrice(double fullPrice) {
        this.fullPrice = fullPrice;
    }

    public double getSalePrice() {
        return salePrice;
    }

    public Double getDiscount() {
        return discount == null ? 0.0 : discount;
    }

    public void setSalePrice(double salePrice) {
        this.salePrice = salePrice;
    }

    public boolean isAvailability() {
        return availability;
    }

    public void setAvailability(boolean availability) {
        this.availability = availability;
    }

    public Supplier getSupplier() {
        return supplier;
    }

    public void setSupplier(Supplier supplier) {
        this.supplier = supplier;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Product product = (Product) o;
        return id == product.id &&
                Double.compare(product.fullPrice, fullPrice) == 0 &&
                Double.compare(product.salePrice, salePrice) == 0 &&
                Double.compare(product.discount, discount) == 0 &&
                availability == product.availability &&
                Objects.equals(productName, product.productName) &&
                Objects.equals(category, product.category) &&
                Objects.equals(supplier, product.supplier);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, productName, category, fullPrice, salePrice, discount, availability, supplier);
    }

    @Override
    public String toString() {
        return "Product{" +
                "id=" + id +
                ", productName='" + productName + '\'' +
                ", category=" + category +
                ", fullPrice=" + fullPrice +
                ", salePrice=" + salePrice +
                ", discount=" + discount +
                ", availability=" + availability +
                ", supplier=" + supplier +
                '}';
    }
}
